doc/src/devdocs/pkgimg.md

```diff
@@ -0,0 +1,48 @@
+# Package Images
+
+Julia package images provide object (native code) caches for Julia packages.
+They are similar to Julia's [system image](@ref dev-sysimg) and support many of the same features.
+In fact the underlying serialization format is the same, and the system image is the base image that the package images are build against.
+
+## High-level overview
+
+Package images are shared libraries that contain both code and data. Like `.ji` cache files, they are generated per package. The data section contains both global data (global variables in the package) as well as the necessary metadata about what methods and types are defined by the package. The code section contains native objects that cache the final output of Julia's LLVM-based compiler.
+
+The command line option `--pkgimage-native-code=no` can be used to turn off object caching for this session. Note that this means that cache files have to likely be regenerated.
+See [`JULIA_MAX_NUM_PRECOMPILE_FILES`](@ref env-max-num-precompile-files) for the upper limit of variants Julia caches per default.
```

Comment:
@vchuravy Thanks for the clarification. However, it is imho still not clear how turning off the use of native package images can cause cache files to be regenerated: IIUC, this means that instead of the `.so` file, only the `.ji` files are used. Or am I missing something here?

Reply:
```
➜  julia rm -rf  ~/.julia/compiled/v1.10/*/*.{ji,so}
➜  julia JULIA_DEBUG=loading ./julia
julia> using SnoopCompileCore
[ Info: Precompiling SnoopCompileCore [e2b509da-e806-4183-be48-004708413034]
┌ Debug: Loading object cache file /home/vchuravy/.julia/compiled/v1.10/SnoopCompileCore/lePIX_UXtTV.so for SnoopCompileCore [e2b509da-e806-4183-be48-004708413034]
└ @ Base loading.jl:947


➜  julia JULIA_DEBUG=loading ./julia --pkgimage-native-code=no
julia> using SnoopCompileCore
┌ Debug: Rejecting cache file /home/vchuravy/.julia/compiled/v1.10/SnoopCompileCore/lePIX_UXtTV.ji for SnoopCompileCore [e2b509da-e806-4183-be48-004708413034] since the flags are mismatched
│   cachefile_flags = 0x23
│   current_flags = 0x22
└ @ Base loading.jl:2513
[ Info: Precompiling SnoopCompileCore [e2b509da-e806-4183-be48-004708413034]
┌ Debug: Loading cache file /home/vchuravy/.julia/compiled/v1.10/SnoopCompileCore/lePIX_NNAS1.ji for SnoopCompileCore [e2b509da-e806-4183-be48-004708413034]
└ @ Base loading.jl:950

➜  ~ ls -lah  /home/vchuravy/.julia/compiled/v1.10/SnoopCompileCore/
-rw-r--r--  1 vchuravy vchuravy 409K Dec 14 15:36 lePIX_NNAS1.ji
-rw-r--r--  1 vchuravy vchuravy  18K Dec 14 15:35 lePIX_UXtTV.ji
-rwxr--r--  1 vchuravy vchuravy 438K Dec 14 15:35 lePIX_UXtTV.so*
```

When we load a package we will check the available cache files and load the most recently created one that is compatible.

Reply:
OK, now I get it. That is, the `.ji` file that goes with the `.so` file is incompatible to the "usual" `.ji` file. Maybe this can be clarified in the docs then, e.g., with something like
```suggestion
The command line option `--pkgimage-native-code=no` can be used to turn off object caching for this session. Note that this means that non-native cache files have to likely be regenerated.
See [`JULIA_MAX_NUM_PRECOMPILE_FILES`](@ref env-max-num-precompile-files) for the upper limit of variants Julia caches per default.
```
?

Reply:
It's a bit odd since technically we could have used the `.so` and only loaded the data section of it and disregard the code section. In fact this is what `--sysimage-native-code=no` means. There is the proposal that instead we should use `--compile=min` to generate non-native cachefiles. 

For me there is the question of what do we want/need to support. Does anyone want to load a `.so` file and only use the data from it and not the code? Does anyone want to limit the generation of native cachefiles, but be happy to use them if they are available? What is implemented right now is a "I want to opt-out of pkgimage creation and loading", but maybe we do want to break it out into different flags.

Functionality wise we need the "do not use native code" option for things like code-coverage and bounds checking.

Reply:
Thanks for the clarifications. I am no expert on the various use cases, where you'd want to turn off selectively only parts of the package image mechanism. However, since object caching is a brand new feature with (possibly) unknown consequences, I'd propose to have a single command line argument that turns off the complete pkgimg machinery and to revert to the current way of precompilation. Any additional fine-grained settings should be on top of that.

Thinking about it, is there an environment variable that has the same effect (I didn't see one)? If not, I propose to add one for, e.g., systems where disk space is limited (user quotas anyone?), such that an affected user could just define sth like `JULIA_PKGIMAGE_NATIVE_CODE=no` in their profile and not have to worry about it in every single invocation of Julia.